tests/integration/schema/migrations/query/with_p2p_schema_branch_test.go

```diff
@@ -0,0 +1,138 @@
+import (
+	"testing"
+
+	"github.com/lens-vm/lens/host-go/config/model"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	testUtils "github.com/sourcenetwork/defradb/tests/integration"
+	"github.com/sourcenetwork/defradb/tests/lenses"
+)
+
+func TestSchemaMigrationQueryWithP2PReplicatedDocOnOtherSchemaBranch(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			testUtils.RandomNetworkingConfig(),
+			testUtils.RandomNetworkingConfig(),
+			testUtils.SchemaUpdate{
+				Schema: `
+					type Users {
+						name: String
+						verified: Boolean
+					}
+				`,
+			},
+			testUtils.SchemaPatch{
+				// Patch first node only
+				NodeID: immutable.Some(0),
+				Patch: `
+					[
+						{ "op": "add", "path": "/Users/Fields/-", "value": {"Name": "email", "Kind": "String"} }
+					]
+				`,
+			},
+			testUtils.ConfigureMigration{
+				// Register the migration on both nodes.
+				LensConfig: client.LensConfig{
+					SourceSchemaVersionID:      "bafkreihax57fohcdupqr2l4heoqxdsiggjfeaubr44tgrz4xqdgvnid4xy",
+					DestinationSchemaVersionID: "bafkreifer354qmdrwdtae5n3k7sbl2oauis3mz24fk46p3otub7ojznobq",
+					Lens: model.Lens{
+						Lenses: []model.LensModule{
+							{
+								Path: lenses.SetDefaultModulePath,
+								Arguments: map[string]any{
+									"dst":   "name",
+									"value": "Fred",
+								},
+							},
+						},
+					},
+				},
+			},
+			testUtils.SchemaPatch{
+				// Patch second node with different patch
+				NodeID: immutable.Some(1),
+				Patch: `
+					[
+						{ "op": "add", "path": "/Users/Fields/-", "value": {"Name": "phone", "Kind": 11} }
+					]
+				`,
+				Lens: immutable.Some(model.Lens{
+					Lenses: []model.LensModule{
+						{
+							Path: lenses.SetDefaultModulePath,
+							Arguments: map[string]any{
+								"dst":   "phone",
+								"value": "1234567890",
+							},
+						},
+					},
+				}),
+			},
+			testUtils.ConfigureReplicator{
+				SourceNodeID: 0,
+				TargetNodeID: 1,
+			},
+			testUtils.CreateDoc{
+				// Create John on the first (source) node only, and allow the value to sync
+				NodeID: immutable.Some(0),
+				Doc: `{
+					"name": "John",
+					"verified": true
+				}`,
+			},
+			testUtils.WaitForSync{},
+			testUtils.Request{
+				// Node 0 should yield results as they were defined, as the newer schema version is
+				// unknown to this node.
+				NodeID: immutable.Some(0),
+				Request: `query {
+					Users {
+						name
+					}
+				}`,
+				Results: []map[string]any{
+					{
+						"name":     "John",
+						"verified": true,
+					},
+				},
+			},
+			testUtils.Request{
+				// Node 1 should yield results migrated down to schema version 1, then up to schema version 3.
+				NodeID: immutable.Some(1),
+				Request: `
+					query {
+						Users {
+							name
+							phone
+						}
+					}
+				`,
+				Results: []map[string]any{
+					{
+						// name has been cleared by the inverse of the migration from version 1 to 2
+						"name": nil,
```

Comment:
thought: This is an unexpected behaviour. In my mind this should have had the value set by the first migration. I'm might be missing why we are doing the inverse migration here.

Reply:
I'm not sure I follow - we defined the inverse/migration action because we want to test it.

The reason it executes is because the document was defined on schema version 2, and to get to schema version 3 we need to inverse to 1, then migrate to 2.  The schema version DAG looks like the below:

```
       (1)
     |     |
   (2)     (3) 
```

Reply:
Ah. I see where I missed. In `TestSchemaMigrationQuery_WithBranchingSchema` we set the active version back to the first one so it was clearer a little bit. Here, each patch is done on a separate node. That's what I missed. All good 👍

Reply:
Note: I'm looking forward to the day this test can revert back to `John` instead of `nil`.

Reply:
> Note: I'm looking forward to the day this test can revert back to John instead of nil.

I don't understand - `nil` is the desired outcome.  We will never go directly from schema version 2 to 3 with a schema history like this - we always need to respect the dag history.

Reply:
I would think the desired outcome is `John` since that is what it was set to on a field that exists on all versions. And didn't need to go through the migration at any point.

Reply:
You raise an interesting point. This implies that migration can be destructive when sharing data across nodes that aren't on the same schema version for a collection if they know of the migration. Setting a migration without the schema involved existing on the node should come with a big warning. I know the DAG isn't affected but the view is and that may cause user to panic when they see (because the view is what they see) their data disappear.

Reply:
> This implies that migration can be destructive when sharing data across nodes that aren't on the same schema version for a collection if they know of the migration

Lens transforms in general can be destructive, there is nothing special/new here.  The lens transform choses to override existing data, that is a choice available to all Lenses, since Lens has existed.  Not warning to this specific flavour here is required, as it is just an inherent fact about Lens.

Reply:
Note: We had a live conversation about this and the conclusion is that Lens migrations can be destructive but that is simply a result of their implementation flexibility and that is reflected in the Lens migration used in this test. 

All good 👍